File: p009.cpp
// Project Euler
// Problem 9: a, b, c are a Pythagorean triplet. If a + b + c = 1000
//            Find the product of abc.

#include <iostream>
using namespace std;

int compute()
{
    int a, b, c;
    for (a = 998; a > 1; --a) {
        for (b = 1; b <= (1000-a)/2; ++b) {
            c = 1000 - a - b;

            if ((a*a) == ((b*b)+(c*c))) break;
        }
        if ((a*a) == ((b*b)+(c*c))) break;
    }

    /*
    cout << "a: " << a << endl;
    cout << "b: " << b << endl;
    cout << "c: " << c << endl;
    */

    return a*b*c;
}

int main()
{
    cout << "The product of abc is." << endl;
    cout << compute() << endl;

    cout << "\nThere is no general case in this problem." << endl;

    return 0;
}

File: Active/Go/solutions/p009.go
// Project Euler
// Problem 9: a, b, c, are a Pythagorean triplet. If a + b + c = 1000
//            Find the product of abc

package go_euler

func Problem_009(sum_of_all int) int {
    // minus 2 because b, c first value are at least 1
	for a := (sum_of_all - 2); a > 1; a-- {
		// (sum_of_all - a) divided by 2 because at that point we are just
		// going to assign value that already own by b to c
		for b := 1; b <= (sum_of_all - a)/2; b++ {
			c := sum_of_all - a - b
            if (a*a) == (b*b) + (c*c) {
                return a * b * c;
			}
        }
    }

    return -1;
}
